src/org/rust/lang/RustSdkType.kt

```diff
@@ -0,0 +1,105 @@
+package org.rust.lang
+
+import com.intellij.execution.configurations.GeneralCommandLine
+import com.intellij.execution.process.CapturingProcessHandler
+import com.intellij.openapi.diagnostic.Logger
+import com.intellij.openapi.projectRoots.SdkAdditionalData
+import com.intellij.openapi.projectRoots.SdkModel
+import com.intellij.openapi.projectRoots.SdkModificator
+import com.intellij.openapi.projectRoots.SdkType
+import com.intellij.openapi.util.SystemInfo
+import org.jdom.Element
+import org.rust.lang.icons.RustIcons
+import java.io.File
+import java.util.concurrent.ExecutionException
+import java.util.regex.Pattern
+
+class RustSdkType : SdkType("Rust SDK") {
+
+    override fun getPresentableName() = name
+
+    override fun getIcon() = RustIcons.FILE
+    override fun getIconForAddAction() = icon
+
+    override fun suggestHomePath(): String? {
+        if (SystemInfo.isMac) {
+            // look for rust installed by homebrew
+            val homebrew = File("/usr/local/Cellar/rust")
+            if (homebrew.exists())
+                return homebrew.absolutePath
+        }
+
+        return null
+    }
+
+    override fun isValidSdkHome(path: String): Boolean {
+        val rustc = getSdkExecutable(path, "rustc")
+        val cargo = getSdkExecutable(path, "cargo")
+        return rustc.canExecute() && cargo.canExecute()
+    }
+
+    override fun adjustSelectedSdkHome(homePath: String?): String? {
+        val file = File(homePath)
+        return when (file.nameWithoutExtension) {
+            "rustc", "cargo" -> file.parentFile.absolutePath
+            "bin"            -> file.parentFile.absolutePath
+            else             -> super.adjustSelectedSdkHome(homePath)
+        }
+    }
+
+    private fun getSdkExecutable(sdkHome: String, command: String): File {
+        return File(File(sdkHome, "bin"), getExecutableFileName(command))
+    }
+
+    private fun getExecutableFileName(executableName: String): String {
+        return if (SystemInfo.isWindows) executableName + ".exe" else executableName
+    }
+
+    override fun suggestSdkName(currentSdkName: String?, sdkHome: String) =
+            getVersionString(sdkHome)?.let { "Rust " + it }
```

Comment:
I feel that even such short stings are better written as `"Rust $it"`, but this is highly subjective